### PR TITLE
Fix sparse matrix nits, and make sparse matrix construction fail if zero(Tv) not defined

### LINKED ISSUE #9917: Sparse matrix edge cases
Ref: https://github.com/JuliaLang/julia/pull/9750#issuecomment-71353419
cc: @ViralBShah 

There are a few edge cases where the behavior of sparse matrices is inconsistent with those of regular matrices:
```julia
julia> sparse([]) # works fine
0x1 sparse matrix with 0 Any entries:
julia> sparse([]') # should work fine, but doesn't
ERROR: MethodError: `zero` has no method matching zero(::Type{Any})
julia> full(sparse([]))
ERROR: MethodError: `zero` has no method matching zero(::Type{Any})

julia> bVec= []; bSparse = sparse(bVec);
julia> bVec[:1] # this error makes sense
ERROR: BoundsError: attempt to access 0-element Array{Any,1} at index [1]
 in getindex at ./array.jl:260
julia> bSparse[:1] # DivideError makes no sense- we aren't performing divison
ERROR: DivideError: integer division error
 in getindex at sparse/sparsematrix.jl:879

julia> bSparse[2:1] # this should really throw an error
0x1 sparse matrix with 0 Any entries:

julia> e=speye(100);
julia> e[-10:10] # should throw an error
21x1 sparse matrix with 1 Float64 entries:
	[12,  1]  =  1.0
```

```julia
julia> versioninfo()
Julia Version 0.4.0-dev+2892
Commit adc1c83 (2015-01-24 19:45 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i7-4650U CPU @ 1.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

### LINKED ISSUE #11032: integer division error but I expected BoundsError
```
Version 0.3.7 (2015-03-23 21:36 UTC)
Official http://julialang.org release
x86_64-linux-gnu

julia> a = zeros(0,0)
0x0 Array{Float64,2}

julia> b = spzeros(0, 0)
0x0 sparse matrix with 0 Float64 entries:

julia> a[0]
ERROR: BoundsError()
 in getindex at array.jl:246

julia> b[0]
ERROR: integer division error
 in getindex at sparse/sparsematrix.jl:766
```

### PR DESCRIPTION
**Pull request by simonster**

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -184,7 +184,9 @@ convert{T}(::Type{AbstractMatrix{T}}, A::SparseMatrixCSC) = convert(SparseMatrix
 convert(::Type{Matrix}, S::SparseMatrixCSC) = full(S)
 
 function full{Tv}(S::SparseMatrixCSC{Tv})
-    A = zeros(Tv, S.m, S.n)
+    # Handle cases where zero(Tv) is not defined but the array is dense.
+    # (Should we really worry about this?)
+    A = length(S) == nnz(S) ? Array(Tv, S.m, S.n) : zeros(Tv, S.m, S.n)
```
Comment (simonster) [inline base/sparse/sparsematrix.jl]: This was one of the issues reported in #9917 so I "fixed" it, but I'm not sure we should even allow creating a SparseMatrix of a type without `zero` defined.

Comment (ViralBShah) [inline base/sparse/sparsematrix.jl]: That seems reasonable.

**Comment by simonster** [comment]
Added an error in the SparseMatrixCSC constructor if the value type has no defined `zero`. That part is potentially breaking and should not be backported.

**Comment by IainNZ** [comment]
Related PR, that I closed without merging: https://github.com/JuliaLang/julia/pull/9325

**Comment by ViralBShah** [comment]
@lindahua Could you give some examples? It would seem that it should be straightforward to provide a `zero(T)` method. We barely scratched some of the ideas of generalizing sparse matrix operations to graph operations here:

http://www.mit.edu/~kepner/pubs/JuliaSemiring_HPEC2013_Paper.pdf

It could very well be that if you care about having some of these things working, you should provide `zero(T)`, but it doesn't need to be mandatory in case you are only doing operations where `zero(T)` is unlikely to get invoked.

**Comment by tkelman** [comment]
I have sparse matrices with symbol or `Expr` or `Any` element types every now and then, where it feels very very wrong to be defining `zero(::Any)` or `zero(::Symbol)` etc if I don't actually have to.

**Comment by ScottPJones** [comment]
Has anybody thought about making this more general?  Instead of looking for `zero()`,
look for an `emptyvalue()` method.  For `T <: Number`, it would still be `zero()`, but for
`T <: AbstractString`, it could be `T("")`, and the generic fallback for `T <: Any`, it would be `Void`.
I think for numeric vectors and matrices, you'd still get exactly what you get now, but this would
make sparse vectors and matrices useable (and efficient) for non-numeric data, without having people do strange things like define `zero(::Any)` or `zero(::AbstractString)`.

**Comment by lindahua** [comment]
This is actually about a more fundamental problem than just the ``zero`` method.

It is about the *concept* of *sparse array* (or *sparse matrix*).  We have two options:

- Consider *sparse arrays* within the numerical context. This is what most of the methods in the Sparse module were designed for. With ``Tv`` being a non-numerical value, you won't be able to use most of these functions (that's not just about the ``zero`` method, they also use ``+`` and ``*``, etc).

Implementing this concept is very simple -- just have ``Tv <: Number`` (you don't have to check whether ``Tv`` implements ``zero`` or not).

- Consider *sparse arrays* as a sorted map from indexes to values, kind of like a ``Dict{NTuple{N, Ti}, Tv}``, but with a more compact internal representation. With this concept, you can put whatever you want as ``Tv``. However, it makes no sense to call any numerical methods on the array (which requires not only ``zero``, but also numerical operators being defined on ``Tv``). 

The only method that I can conceive that may be made to work for such arrays is ``full`` (by adding ``zero`` or ``emptyvalue``, etc). Actually, there is a more flexible way to make this work:

```julia
function full(A::SparseMatrixCSC{T}, emptyval::T)
# ...
end

full{T<:Number}(A::SparseMatrixCSC{T}) = full(A, zero(T));
```

You don't need to define ``emptyvalue`` for everything, which, IMO, is not really feasible. There are plenty of types, for which it is unclear what is an empty value. So if you want to call ``full`` on a non-numerical sparse array, just supply what you think is a proper emptyvalue depending on your context.

I guess @ViralBShah might be interested in using a sparse matrix as a graph. In those cases, we don't need to have ``zero`` at all. The general algorithmic pattern for graph theoretical algorithms is as follows:

```julia
for s in vertices(g)
     # ...
     for t in neighbors(g, s)  
     end
end 
```
For such type of algorithms, what you really need is just a way to quickly give you the list of *nonzeros* of a certain column (or row), and you don't need to bother with those that are not stored. (With the CSC format, this can be efficiently done).

**Comment by lindahua** [comment]
It is very important to note that the primary purpose of sparse storage is to make it possible to have algorithms with complexity ``O(nnz)`` instead of ``O(n)``.  Scanning over all elements (stored or not) defeats this purpose.

The only method that needs to scan all elements is ``full``, which I already provided a solution that does not require a ``zero`` method on ``Tv``.

**Comment by ScottPJones** [comment]
For non-numerical applications, I'd think the primary purpose of sparse storage is to save memory,  accepting O(log n) (n = num not null) time instead of O(1) to find a particular field in a row...

**Comment by ViralBShah** [comment]
Would you not want to use a hash table in such cases?

**Comment by tkelman** [comment]
> The only method that I can conceive that may be made to work for such arrays is `full`

Indexing and traversal of the stored values works fine too. There are several contexts where I'm putting objects in a sparse matrix that are numberlike containers but not `<: Number`, or are functions that evaluate to numbers but are not themselves numbers (Jacobian and Hessian matrices, etc). These will inevitably not work as completely as `Tv <: Number` will, but they're useful nonetheless.

**Comment by simonster** [comment]
So it sounds like we don't want to restrict SparseMatrixCSC to types with `zero`, at least for the moment. Shall I go ahead and merge just the first commit here?

**Comment by tkelman** [comment]
2fedc274232bdf944dcc86eb05265852de7d1262 looks fine and non-controversial

**Comment by tkelman** [comment]
backported in c8804e8c47b27ab66b0a9649a408c0f47233464e

**Pull request merged**